impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultDependencyResolverResult.java

```diff
@@ -102,19 +102,27 @@ public class DefaultDependencyResolverResult implements DependencyResolverResult
      * to {@link #addDependency(Node, Dependency, Predicate, Path)}.
      *
      * @param request the corresponding request
-     * @param cache cache of module information about each dependency
      * @param exceptions the exceptions that occurred while building the dependency graph
      * @param root the root node of the dependency graph
      * @param count estimated number of dependencies
      */
     public DefaultDependencyResolverResult(
+            DependencyResolverRequest request, List<Exception> exceptions, Node root, int count) {
+        this(request, new PathModularizationCache(), exceptions, root, count);
+    }
+
+    DefaultDependencyResolverResult(
             DependencyResolverRequest request,
             PathModularizationCache cache,
             List<Exception> exceptions,
             Node root,
             int count) {
         this.request = request;
-        this.cache = cache;
+        if (cache == null) {
```

Comment:
This `if` … `else` block could be replaced by `this.cache = Objects.requireNonNull(cache)`. For consistency, we may do the same for other arguments too except `root` which is nullable.

Reply:
I don't think that's the same. cache is nullable. We need to create a cache if the argument is null, not throw an exception. Null is passed for this argument and we don't want to break those existing usages.

Reply:
The new constructor should replace all the calls to the previous constructor with a null cache. We are sure that the new constructor is invoked outside the package since the old constructor became package-private. So we only need to ensure that the calls inside the package pass a non-null cache, and I think that it is already the case.

Reply:
It's risky and brittle to allow null caches given the multiple locations that invoke methods on the cache and the lack of test coverage on this class. It is unsafe to allow this field to be null, even if the constructor argument is. Even if it doesn't cause an uncaught NPE now it will in the future.

Reply:
Yes I agree, which is why my proposal above was to use `Objects.requireNonNull`.

Reply:
The rational was that a cache created here is of limited use. The intend was to have a session-wide cache, which is why this method was requesting the cache in argument. The creation of a cache in the constructor was only a temporary workaround for the fact that we have not investigated where a session-wide cache should be stored.

Reply:
Whatever the intent was, this field is dereferenced without null checks in multiple places. That's dangerous.